cluster/cluster.go

```diff
@@ -15,4 +38,85 @@ func Init(name, version string, primary bool, started time.Time) {
 		primaryChange: time.Now(),
 		stateChange:   time.Now(),
 	}
+	Mode = m
+	go probePeers(interval)
+}
+
+func Stop() {
+	close(shutdown)
+}
+
+func AddPeer(remoteAddr *url.URL) {
+	mu.Lock()
+	peer := &Node{
+		remoteAddr:  remoteAddr,
+		stateChange: time.Now(),
+	}
+	peers = append(peers, peer)
+	go peer.Probe()
+	mu.Unlock()
+}
+
+func GetPeers() []*Node {
+	mu.Lock()
+	p := make([]*Node, len(peers))
+	copy(p, peers)
+	mu.Unlock()
+	return p
+}
+
+func probePeers(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for {
+		select {
+		case <-shutdown:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			mu.Lock()
+			for _, peer := range peers {
+				go peer.Probe()
+			}
+			mu.Unlock()
+		}
+	}
+}
+
+// return the list of peers to broadcast requests to
+// Only 1 peer per partition is returned
+func PeersForQuery() []*Node {
+	// If we are running in single mode, just return thisNode
+	if Mode == ModeSingle {
+		return []*Node{ThisNode}
+	}
+
+	peersMap := make(map[int32][]*Node)
+	for _, part := range ThisNode.GetPartitions() {
+		peersMap[part] = []*Node{ThisNode}
+	}
+	mu.Lock()
+	for _, peer := range peers {
+		if !peer.IsReady() {
+			continue
+		}
+		for _, part := range peer.GetPartitions() {
+			peersMap[part] = append(peersMap[part], peer)
+		}
+	}
+	mu.Unlock()
+	selectedPeers := make(map[string]*Node)
+	answer := make([]*Node, 0)
+	// we only need to query 1 node for each partition.
+	// we are just using the first node for each partition, which will be
+	// ThisNode if ThisNode owns the partition.  Later we can add more
+	// complex scheduling logic to distribute queries across the cluster
```

Comment:
I still have to review the changes in the `api` package, but one important concern that comes up with my partition assignment approach (which is not an issue with yours) is that when we query multiple nodes and there is some overlap in between, they might do duplicate work, unless we specify in the request which partitions we want to use the node for.

for example: let's say ThisNode is A, and we query data for all 4 partitions.
we'll pick A for 0 and 1. let's say we select B for 2 and D for 3. if we send the request without further specification, then D will respond with data for both part 2 and 3 because it has data for it, even though we selected B to do the work for partition 2.